lib/irb/cmd/debug.rb

```diff
@@ -5,28 +5,90 @@ module IRB
-          end
-          frames
+        unless binding_irb?
+          puts "`debug` command is supported only on binding.irb"
+          return
+        end
+
+        unless setup_debugger
+          puts <<~MSG
+            You need to install the debug gem before using this command.
+            If you use `bundle exec`, please add `gem "debug"` into your Gemfile.
+          MSG
+          return
         end
 
+        # To make debugger commands like `next` or `continue` work without asking
+        # the user to quit IRB after that, we need to exit IRB first and then hit
+        # a TracePoint on #debug_break.
         file, lineno = IRB::Irb.instance_method(:debug_break).source_location
         DEBUGGER__::SESSION.add_line_breakpoint(file, lineno + 1, oneshot: true, hook_call: false)
         # exit current Irb#run call
         throw :IRB_EXIT
-      rescue LoadError => e
-        puts <<~MSG
-          You need to install the debug gem before using this command.
-          If you use `bundle exec`, please add `gem "debug"` into your Gemfile.
-        MSG
+      end
+
+      private
+
+      def binding_irb?
+        caller.any? do |frame|
+          BINDING_IRB_FRAME_REGEXPS.any? do |regexp|
+            frame.match?(regexp)
+          end
+        end
+      end
+
+      def setup_debugger
+        unless defined?(DEBUGGER__::SESSION)
+          begin
+            require "debug/session"
+          rescue LoadError # debug.gem is not written in Gemfile
+            return false unless load_bundled_debug_gem
+          end
+          DEBUGGER__.start(nonstop: true)
+        end
+
+        unless DEBUGGER__.respond_to?(:capture_frames_without_irb)
+          DEBUGGER__.singleton_class.send(:alias_method, :capture_frames_without_irb, :capture_frames)
+
+          def DEBUGGER__.capture_frames(*args)
+            frames = capture_frames_without_irb(*args)
+            frames.reject! do |frame|
+              frame.realpath&.start_with?(IRB_DIR) || frame.path == "<internal:prelude>"
+            end
+            frames
+          end
+        end
+
+        true
+      end
+
+      # This is used when debug.gem is not written in Gemfile. Even if it's not
+      # installed by `bundle install`, debug.gem is installed by default because
+      # it's a bundled gem. This method tries to activate and load that.
+      def load_bundled_debug_gem
+        # Discover debug.gem in GEM_HOME
+        debug_gem = Dir.glob("#{Gem.paths.home}/gems/debug-*").sort.reverse.find do |path|
+          File.basename(path).match?(/\Adebug-\d+\.\d+\.\d+\z/)
+        end
+        return false unless debug_gem
+
+        # Attempt to forcibly load the bundled gem
+        $LOAD_PATH << "#{debug_gem}/lib"
+        begin
+          require "debug/session"
+        rescue LoadError
+          false
+        else
+          puts "Loaded #{File.basename(debug_gem)}"
```

Comment:
Can these 2 lines be put under `require "debug/session"` so we don't need the `else`?

Reply:
I agree with your argument and that's why I didn't made the same argument for the similar rescue block in `setup_debugger`.
So the difference between our views is actually how much of the chance this puts can cause the error we're trying to rescue. From my personal understanding on `File.basename` and `puts`, I thought it'd be very small so it's worth moving it here.